baseapp/baseapp.go

```diff
@@ -842,8 +847,8 @@ func (app *BaseApp) runCheckTxConcurrently(mode execMode, txBytes []byte) (gInfo
 	// embedded here to ensure that the lifetime of the mutex is limited to only this function allowing
 	// for the return values to be computed without holding the lock.
 	func() {
-		app.mtx.Lock()
-		defer app.mtx.Unlock()
+		app.mtx.RLock()
```

Comment:
Why are we confident that mutation after `cacheTxContext` is either safe or not done?

Reply:
We are not. This PR changes the contract so that the AnteHandler must ensure concurrency correctness and why I added a mutex lock AnteDecorator at the first one so that the existing concurrency model was enforced (i.e. a single global lock).